src/python/pants/engine/rules.py

```diff
@@ -28,123 +42,96 @@ class Rule(AbstractClass):
   """
 
   @abstractproperty
-  def input_selectors(self):
-    """Collection of input selectors"""
+  def output_constraint(self):
+    """An output Constraint type for the rule."""
 
   @abstractproperty
-  def func(self):
-    """Rule function."""
+  def input_selectors(self):
+    """Collection of input selectors."""
 
-  @abstractproperty
-  def output_product_type(self):
-    """The product type produced by this rule."""
 
-  def as_triple(self):
-    """Constructs an (output, input, func) triple for this rule."""
-    return (self.output_product_type, self.input_selectors, self.func)
+class TaskRule(datatype('TaskRule', ['output_constraint', 'input_selectors', 'func']), Rule):
+  """A Rule that runs a task function when all of its input selectors are satisfied."""
 
+  def __new__(cls, output_type, input_selectors, func):
+    # Validate result type.
+    if isinstance(output_type, Exactly):
+      constraint = output_type
+    elif isinstance(output_type, type):
+      constraint = Exactly(output_type)
+    else:
+      raise TypeError("Expected an output_type for rule {}, got: {}".format(func, output_type))
 
-class TaskRule(datatype('TaskRule', ['input_selectors', 'func', 'product_type', 'constraint']),
-               Rule):
-  """A Rule that runs a task function when all of its input selectors are satisfied."""
+    # Validate selectors.
+    if not isinstance(input_selectors, (list, tuple)):
```

Comment:
also wonder if it would make sense to require that this be a tuple in all cases? afaict, there's no benefit to making these `list` by default other than that `[]` sticks out more in a call signature than `()`. whatever we do here in the common case is likely to get cargo culted forward.

maybe this could be a named type/tuple type alias? which could then be used in lieu of a kwarg in the decorator, e.g.

```
@produces(Product, WithInputs(SelectThing(), SelectThing()))
```

or something along those lines.